File: 4 - Clustering/4a - K-Means Clustering/kmeans.py
# K-Means Clustering 

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Mall_Customers.csv')
X = dataset.iloc[:, -2:].values

# Using the elbow method to find optimal number of clusters
from sklearn.cluster import KMeans
WCSS = []
for i in range(1,11):
    kmeans = KMeans(n_clusters = i,
                    init = 'k-means++',
                    max_iter = 300,
                    n_init = 10,
                    random_state = 0)
    kmeans.fit(X)
    WCSS.append(kmeans.inertia_)

plt.plot(range(1,11),WCSS)
plt.title('The Elbow Method')
plt.xlabel('Number of clusters')
plt.ylabel('WCSS')
plt.show()

# Applying K-means to the dataset
kmeans = KMeans(n_clusters = 5,
                    init = 'k-means++',
                    max_iter = 300,
                    n_init = 10,
                    random_state = 0)
y_kmeans = kmeans.fit_predict(X)

# Visualising the results
plt.scatter(X[y_kmeans == 0, 0],X[y_kmeans == 0, 1], s = 50, c = 'red', label = 'Careful')
plt.scatter(X[y_kmeans == 1, 0],X[y_kmeans == 1, 1], s = 50, c = 'blue', label = 'Standard')
plt.scatter(X[y_kmeans == 2, 0],X[y_kmeans == 2, 1], s = 50, c = 'yellow', label = 'Target')
plt.scatter(X[y_kmeans == 3, 0],X[y_kmeans == 3, 1], s = 50, c = 'magenta', label = 'Careless')
plt.scatter(X[y_kmeans == 4, 0],X[y_kmeans == 4, 1], s = 50, c = 'cyan', label = 'Sensible')
plt.scatter(kmeans.cluster_centers_[:,0],kmeans.cluster_centers_[:,1], marker='*', s = 300, c = 'black', label = 'Centroids')
plt.title('Cluster of Clients')
plt.xlabel('Annual Income (k$)')
plt.ylabel('Spending Score (1-100)')
plt.legend() 
plt.show()

File: 4 - Clustering/4a - K-Means Clustering/kmeans.R
# K-means Clustering

# Importing the dataset
dataset <- read.csv('Mall_Customers.csv')
X <- dataset[4:5]

# Using the elbow method to find the optimal number of clusters
set.seed(6)
WCSS <- vector()
for (i in 1:10) WCSS[i] <- sum(kmeans(X,i)$withinss)
plot(1:10, WCSS, type = 'b', 
     main = paste('The Elbow Method'),
     xlab = 'Number of clusters',
     ylab = 'WCSS')

# Applying k-means to dataset
set.seed(29)
kmeans <- kmeans(X,5, iter.max = 300,nstart = 10)
y_kmeans = kmeans$cluster

# Visualising the results
library(cluster)
clusplot(X,
         y_kmeans,
         lines = 0,
         shade = TRUE,
         color = TRUE,
         labels = 2,
         plotchar = FALSE,
         span = TRUE,
         main = paste('Clusters of clients'),
         xlab = 'Annual Income',
         ylab = 'Spending Score')
